### ISSUE TITLE
VS 2008 for amd64 is no longer detected

### ISSUE THREAD

**Issue by bdbaddog** [SCons/scons#2965]
This issue was originally created at: 2014-07-21 15:43:04.
This issue was reported by: `kayhayen`.
kayhayen said at 2014-07-21 15:43:04
>In my Python compiler Nuitka, I specify the target architecture, as detected by 
>another Python running potentially under Python3, potentially on another 
>architecture. That is why I do TARGET_ARCH assignment when created the 
>environment.
> 
>For Python2 though, it's running Scons with sys.executable, as that will be good 
>enough for Scons. But it could well be imagined, and happens, that a Python3 
>compilation calls Scons with a Python2 of another architecture.
> 
>However, I noticed that with Scons 2.3.2, only the 32 bit compiler is found, but 
>when TARGET_ARCH is "x86_64", on the same system, that 2.3.0 finds it, it is not 
>found.
> 
>There is something in the release notes about TARGET_ARCH no longer defaulting 
>to HOST_ARCH. Not sure, if that is related.
> 
>So here is 2.3.2 output:
> 
```
$ /c/Python27_32/python bin/nuitka --run --show-scons tests/basics/Asserts.py Scons command: c:\Python27_32\python.exe .\nuitka\build\inline_copy\bin\scons.py  -f .\nuitka\build\SingleExe.scons --jobs 6 --warn=no-deprecated --no-site-dir --debug=explain show_scons=true python_version=2.7 unstriped_mode=false full_compat=true debug_mode=false target_arch=x86 optimize_mode=true source_dir=Asserts.build cache_mode=true name=Asserts python_debug=false python_prefix=c:\Python27_32 module_count=1 result_name=Asserts nuitka_src=.\nuitka\build module_mode=false experimental=false
scons: Reading SConscript files ...
AssertionError: C:\Program Files (x86)\Microsoft Visual Studio 9.0\VC\BIN\cl.exe
:
  File "C:\Users\hayen\repos\Py2C\nuitka\build\SingleExe.scons", line 205:
    assert False, getExecutablePath(env["CXX"], initial = False)
 
hayen@HAYEN-PC ~/repos/Py2C (master)
$ /c/Python27_64/python bin/nuitka --run --show-scons tests/basics/Asserts.py
Scons command: c:\Python27_64\python.exe .\nuitka\build\inline_copy\bin\scons.py
 -f .\nuitka\build\SingleExe.scons --jobs 6 --warn=no-deprecated --no-site-dir -
-debug=explain show_scons=true python_version=2.7 unstriped_mode=false full_comp
at=true debug_mode=false target_arch=x86_64 optimize_mode=true source_dir=Assert
s.build cache_mode=true name=Asserts python_debug=false python_prefix=c:\Python2
7_64 module_count=1 result_name=Asserts nuitka_src=.\nuitka\build module_mode=fa
lse experimental=false
scons: Reading SConscript files ...
AssertionError: None:
  File "C:\Users\hayen\repos\Py2C\nuitka\build\SingleExe.scons", line 205:
    assert False, getExecutablePath(env["CXX"], initial = False)
```
> 
>I am merely dumping what env["CXX"] resolves to. No MSVC is in PATH of course.
> 
>For 2.3.0, this looks like this:
> 
> 
```
>$ /c/Python27_32/python bin/nuitka --run --show-scons tests/basics/Asserts.py
>Scons command: c:\Python27_32\python.exe .\nuitka\build\inline_copy\bin\scons.py
> -f .\nuitka\build\SingleExe.scons --jobs 6 --warn=no-deprecated --no-site-dir -
>-debug=explain show_scons=true python_version=2.7 unstriped_mode=false full_comp
>at=true debug_mode=false target_arch=x86 optimize_mode=true source_dir=Asserts.b
>uild cache_mode=true name=Asserts python_debug=false python_prefix=c:\Python27_3
>2 module_count=1 result_name=Asserts nuitka_src=.\nuitka\build module_mode=false
> experimental=false
>scons: Reading SConscript files ...
>AssertionError: C:\Program Files (x86)\Microsoft Visual Studio 9.0\VC\BIN\cl.exe
>:
>  File "C:\Users\hayen\repos\Py2C\nuitka\build\SingleExe.scons", line 205:
>    assert False, getExecutablePath(env["CXX"], initial = False)
> 
>$ /c/Python27_64/python bin/nuitka --run --show-scons tests/basics/Asserts.py
>Scons command: c:\Python27_64\python.exe .\nuitka\build\inline_copy\bin\scons.py
> -f .\nuitka\build\SingleExe.scons --jobs 6 --warn=no-deprecated --no-site-dir -
>-debug=explain show_scons=true python_version=2.7 unstriped_mode=false full_comp
>at=true debug_mode=false target_arch=x86_64 optimize_mode=true source_dir=Assert
>s.build cache_mode=true name=Asserts python_debug=false python_prefix=c:\Python2
>7_64 module_count=1 result_name=Asserts nuitka_src=.\nuitka\build module_mode=fa
>lse experimental=false
>scons: Reading SConscript files ...
>AssertionError: C:\Program Files (x86)\Microsoft Visual Studio 9.0\VC\BIN\amd64\
>cl.exe:
>  File "C:\Users\hayen\repos\Py2C\nuitka\build\SingleExe.scons", line 205:
>    assert False, getExecutablePath(env["CXX"], initial = False)
> 
```
>Any idea what is going on there?

dirkbaechle said at 2014-07-22 00:16:10
>Kay,
> 
>if you're interested in immediate feedback to your problem, I can only recommend 
>to repost your questions on the User Mailing list at scons-users@scons.org .
> 
>You'll reach a much larger audience there, we have quite a few Windows experts 
>that don't check on this issue database regularly.
> 
>In general, this is the preferred method anyway: Asking on the User ML first, to 
>get confirmed that it's a bug. Then create an issue report for it.
> 
>Thanks a lot in advance for your consideration.
> 
>Dirk

kayhayen said at 2014-07-27 11:28:41
> 
>I was looking at the code diffs, and trying with debug setting as advised on the 
>mailing list. No response to this, I guess, that is more development stuff:
> 
```
-    # If the user hasn't specifically requested a TARGET_ARCH, and
-    # The TARGET_ARCH is amd64 then also try 32 bits if there are no viable
-    # 64 bit tools installed
     try_target_archs = [target_platform]
-    if not req_target_platform and target_platform in ('amd64','x86_64'):
+    debug("msvs_find_valid_batch_script(): req_target_platform %s target_platfo
+
+    # VS2012 has a "cross compile" environment to build 64 bit
+    # with x86_amd64 as the argument to the batch setup script
+    if req_target_platform in ('amd64','x86_64'):
+        try_target_archs.append('x86_amd64')
+    elif not req_target_platform and target_platform in ['amd64','x86_64']:
+        # There may not be "native" amd64, but maybe "cross" x86_amd64 tools
+        try_target_archs.append('x86_amd64')
+        # If the user hasn't specifically requested a TARGET_ARCH, and
+        # The TARGET_ARCH is amd64 then also try 32 bits if there are no viable
+        # 64 bit tools installed
         try_target_archs.append('x86')
```
> 
>This may well explain my trouble. I do request TARGET_ARCH "amd64", but I only 
>have a cross compiler, Visual Studio 2008 installed (required to link against 
>Python2.7). Also, people use 32 and 64 bits Python, and I need to insist the 
>compiler matches.
> 
>So the old code will give "try_target_archs" as [ "x86_64", "x86" ], where the 
>new code only gives me [ "x86_64", "x86_amd64" ], which is more correct.
> 
>However, in my case, "vcvars64.bat" of the "x86" compiler is not attempted in 
>this case. Only "vcallvars.bat x86_amd64", which while listed in the /h output, 
>gives an error when executed. With 2.3.0, this error also occured, but then 
>"vcvars64.bat" was tried, and works.
> 
>So, I believe this is a regression of Scons with VS 2008 and x86_64 as a 
>requested target platform.

kayhayen said at 2014-08-10 11:43:28
>So, is there any progress to hope for with this issue. I kind of nailed it pretty 
>well down, but I suspect nobody really cares anymore, right?
> 
>So, I am therefore requiring my users now to upgrade their MSVC installation. It 
>makes no sense to support 2008 for me anymore, mostly because it cannot be 
>installed by new users anyway. Whatever problem there would be to not using it, 
>other solutions need to be found anyway.
> 
>So I am open to closing this bug as "won't fix" or whatever it is in this bug 
>tracker.

**Comment by mwichmann** [comment]
This topic is addressed in #3538 - setting TARGET_ARCH defeats some of the logic of the "cross" environments, but not sure it will actually apply to MSVC versions this old.  Is there any reason to keep this issue open given the age of the MSVC and the notes in the last comment (which date to 2014)?

**Comment by bdbaddog** [comment]
We can't reproduce this currently as VS2008's not available.
If this is still affecting anyone, please reopen and perhaps pop on the users mailing list, IRC, or discord servers so we can work with you to help provide a fix and have you test it.
